src/elements/element.line.js

```diff
@@ -52,57 +33,104 @@ function setStyle(ctx, vm) {
 	ctx.strokeStyle = vm.borderColor;
 }
 
-function normalPath(ctx, points, spanGaps, vm) {
-	const steppedLine = vm.steppedLine;
-	const lineMethod = steppedLine ? helpers.canvas._steppedLineTo : helpers.canvas._bezierCurveTo;
-	let move = true;
-	let index, currentVM, previousVM;
+function lineTo(ctx, previous, target) {
+	ctx.lineTo(target.x, target.y);
+}
+
+function getLineMethod(line) {
+	if (line.steppedLine) {
+		return _steppedLineTo;
+	}
+
+	if (line.tension) {
+		return _bezierCurveTo;
+	}
+
+	return lineTo;
+}
+
+function getInterpolationMethod(line) {
+	if (line.steppedLine) {
+		return _steppedInterpolation;
+	}
 
-	for (index = 0; index < points.length; ++index) {
-		currentVM = points[index]._view;
+	if (line.tension) {
+		return _bezierInterpolation;
+	}
 
-		if (currentVM.skip) {
-			move = move || !spanGaps;
+	return _pointInLine;
+}
+
+function pathSegment(ctx, line, segment, options) {
+	const {start, end, loop} = segment;
+	const {points, _view: vm} = line;
+	const lineMethod = getLineMethod(vm);
+	const count = points.length;
+	let {move = true, reverse} = options || {};
+	let ilen = end < start ? count + end - start : end - start;
+	let i, point, prev;
+
+	for (i = 0; i <= ilen; ++i) {
+		point = points[(start + (reverse ? ilen - i : i)) % count]._view;
+
+		if (point.skip) {
+			// If there is a skipped point inside a segment, spanGaps must be true
 			continue;
-		}
-		if (move) {
-			ctx.moveTo(currentVM.x, currentVM.y);
+		} else if (move) {
+			ctx.moveTo(point.x, point.y);
 			move = false;
-		} else if (vm.tension || steppedLine) {
-			lineMethod(ctx, previousVM, currentVM, false, steppedLine);
 		} else {
-			ctx.lineTo(currentVM.x, currentVM.y);
+			lineMethod(ctx, prev, point, reverse, vm.steppedLine);
```

Comment:
Do we have to pass `vm.steppedLine` here? Won't `lineMethod` already be using the stepped version? (i.e. `_steppedLineTo`)

Reply:
Its the `mode` for steppedLineTo (before, middle or after)